File: other-programming-languages/Typescript/2024/2024-010/2024-010.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var INPUT_FILE = '\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-010\\input.txt';
function readMap(filename) {
    var lines = fs.readFileSync(filename, 'utf-8').split('\n').filter(function (line) { return line.trim() !== ''; });
    return lines.map(function (line) { return line.trim().split('').map(function (ch) { return parseInt(ch, 10); }); });
}
function neighbors(r, c, rows, cols) {
    var directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    return directions
        .map(function (_a) {
        var dr = _a[0], dc = _a[1];
        return [r + dr, c + dc];
    })
        .filter(function (_a) {
        var nr = _a[0], nc = _a[1];
        return nr >= 0 && nr < rows && nc >= 0 && nc < cols;
    });
}
function findTrailheadScores(grid) {
    var rows = grid.length;
    var cols = grid[0].length;
    var trailheads = [];
    for (var r = 0; r < rows; r++) {
        for (var c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }
    var totalScore = 0;
    for (var _i = 0, trailheads_1 = trailheads; _i < trailheads_1.length; _i++) {
        var _a = trailheads_1[_i], startR = _a[0], startC = _a[1];
        var visited = new Set();
        var queue = [];
        var reachableNines = new Set();
        queue.push([startR, startC]);
        visited.add("".concat(startR, ",").concat(startC));
        while (queue.length > 0) {
            var _b = queue.shift(), r = _b[0], c = _b[1];
            var currentHeight = grid[r][c];
            if (currentHeight === 9) {
                reachableNines.add("".concat(r, ",").concat(c));
            }
            else {
                var nextHeight = currentHeight + 1;
                for (var _c = 0, _d = neighbors(r, c, rows, cols); _c < _d.length; _c++) {
                    var _e = _d[_c], nr = _e[0], nc = _e[1];
                    if (!visited.has("".concat(nr, ",").concat(nc)) && grid[nr][nc] === nextHeight) {
                        visited.add("".concat(nr, ",").concat(nc));
                        queue.push([nr, nc]);
                    }
                }
            }
        }
        totalScore += reachableNines.size;
    }
    return totalScore;
}
function countPaths(r, c, grid, dp, rows, cols) {
    if (dp[r][c] !== null) {
        return dp[r][c];
    }
    var currentHeight = grid[r][c];
    if (currentHeight === 9) {
        dp[r][c] = 1;
        return 1;
    }
    var totalPaths = 0;
    var nextHeight = currentHeight + 1;
    for (var _i = 0, _a = neighbors(r, c, rows, cols); _i < _a.length; _i++) {
        var _b = _a[_i], nr = _b[0], nc = _b[1];
        if (grid[nr][nc] === nextHeight) {
            totalPaths += countPaths(nr, nc, grid, dp, rows, cols);
        }
    }
    dp[r][c] = totalPaths;
    return totalPaths;
}
function calculateTotalRating(grid) {
    var rows = grid.length;
    var cols = grid[0].length;
    var trailheads = [];
    for (var r = 0; r < rows; r++) {
        for (var c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }
    var dp = Array.from({ length: rows }, function () { return Array(cols).fill(null); });
    return trailheads.reduce(function (total, _a) {
        var tr = _a[0], tc = _a[1];
        return total + countPaths(tr, tc, grid, dp, rows, cols);
    }, 0);
}
function main() {
    var startTimePart1 = performance.now();
    var grid = readMap(INPUT_FILE);
    var totalScore = findTrailheadScores(grid);
    var endTimePart1 = performance.now();
    console.log("Part 1 Result: ".concat(totalScore));
    console.log("Time taken for Part 1: ".concat((endTimePart1 - startTimePart1) / 1000, "s"));
    var startTimePart2 = performance.now();
    var totalRating = calculateTotalRating(grid);
    var endTimePart2 = performance.now();
    console.log("Part 2 Result: ".concat(totalRating));
    console.log("Time taken for Part 2: ".concat((endTimePart2 - startTimePart2) / 1000, "s"));
}
main();

File: other-programming-languages/Rust-003/2024/2024-010/2024-010.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;

const INPUT_FILE: &str = "input.txt";

fn read_map(filename: &str) -> Vec<Vec<i32>> {
    let mut grid = Vec::new();
    if let Ok(lines) = read_lines(filename) {
        for line in lines {
            if let Ok(line) = line {
                let row: Vec<i32> = line
                    .trim()
                    .chars()
                    .filter_map(|ch| ch.to_digit(10).map(|d| d as i32))
                    .collect();
                grid.push(row);
            }
        }
    }
    grid
}

fn neighbors(r: usize, c: usize, rows: usize, cols: usize) -> Vec<(usize, usize)> {
    let directions = [(-1, 0), (1, 0), (0, -1), (0, 1)];
    directions
        .iter()
        .filter_map(|&(dr, dc)| {
            let nr = r as isize + dr;
            let nc = c as isize + dc;
            if nr >= 0 && nr < rows as isize && nc >= 0 && nc < cols as isize {
                Some((nr as usize, nc as usize))
            } else {
                None
            }
        })
        .collect()
}

fn find_trailhead_scores(grid: &[Vec<i32>]) -> i32 {
    let rows = grid.len();
    let cols = grid[0].len();

    let mut trailheads = Vec::new();
    for r in 0..rows {
        for c in 0..cols {
            if grid[r][c] == 0 {
                trailheads.push((r, c));
            }
        }
    }

    let mut total_score = 0;

    for (start_r, start_c) in trailheads {
        let mut visited = HashSet::new();
        let mut queue = VecDeque::new();
        let mut reachable_nines = HashSet::new();

        queue.push_back((start_r, start_c));
        visited.insert((start_r, start_c));

        while let Some((r, c)) = queue.pop_front() {
            let current_height = grid[r][c];

            if current_height == 9 {
                reachable_nines.insert((r, c));
            } else {
                let next_height = current_height + 1;
                for (nr, nc) in neighbors(r, c, rows, cols) {
                    if !visited.contains(&(nr, nc)) && grid[nr][nc] == next_height {
                        visited.insert((nr, nc));
                        queue.push_back((nr, nc));
                    }
                }
            }
        }

        total_score += reachable_nines.len() as i32;
    }

    total_score
}

fn count_paths(
    r: usize,
    c: usize,
    grid: &[Vec<i32>],
    dp: &mut Vec<Vec<Option<i32>>>,
    rows: usize,
    cols: usize,
) -> i32 {
    if let Some(value) = dp[r][c] {
        return value;
    }

    let current_height = grid[r][c];

    if current_height == 9 {
        dp[r][c] = Some(1);
        return 1;
    }

    let next_height = current_height + 1;
    let mut total_paths = 0;

    for (nr, nc) in neighbors(r, c, rows, cols) {
        if grid[nr][nc] == next_height {
            total_paths += count_paths(nr, nc, grid, dp, rows, cols);
        }
    }

    dp[r][c] = Some(total_paths);
    total_paths
}

fn calculate_total_rating(grid: &[Vec<i32>]) -> i32 {
    let rows = grid.len();
    let cols = grid[0].len();

    let mut trailheads = Vec::new();
    for r in 0..rows {
        for c in 0..cols {
            if grid[r][c] == 0 {
                trailheads.push((r, c));
            }
        }
    }

    let mut dp = vec![vec![None; cols]; rows];

    let mut total_rating = 0;
    for (tr, tc) in trailheads {
        total_rating += count_paths(tr, tc, grid, &mut dp, rows, cols);
    }

    total_rating
}

fn main() {
    let start_part1 = Instant::now();
    let grid = read_map(INPUT_FILE);
    let total_score = find_trailhead_scores(&grid);
    let duration_part1 = start_part1.elapsed();

    println!("Part 1 Result: {}", total_score);
    println!("Time taken for Part 1: {:.9} s", duration_part1.as_secs_f64());

    let start_part2 = Instant::now();
    let total_rating = calculate_total_rating(&grid);
    let duration_part2 = start_part2.elapsed();

    println!("Part 2 Result: {}", total_rating);
    println!("Time taken for Part 2: {:.9} s", duration_part2.as_secs_f64());
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
